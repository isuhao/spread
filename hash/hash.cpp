#include "hash.hpp"

using namespace Spread;

#include "../libs/sha2/sha2.h"
#include <string.h>
#include <stdexcept>

void Hash::hash(const void *input, uint32_t len)
{
  dealloc();
  sha256((const unsigned char*)input, len, data);
  size() = len;
}

void Hash::update(const void *input, uint32_t len)
{
  // Create a new context if this is the first call
  if(context == NULL)
    {
      clear();
      context = new sha256_ctx;
      sha256_init((sha256_ctx*)context);
      size() = 0;
    }

  // Update the context with the new data
  sha256_update((sha256_ctx*)context, (const unsigned char*)input, len);
  size() += len;
}

Hash Hash::finish()
{
  /* If finish is called without update() being called first, we need
     to treat it as if we are hashing a zero-length buffer.

     Finish may be preceded by any number of update() calls, including
     zero.

     Note that the hash of a zero-length buffer is NOT the same as an
     empty (zero) hash!
   */
  if(!context)
    hash(NULL, 0);
  else
    {
      sha256_ctx *ctx = (sha256_ctx*)context;

      // Finish up and delete the context
      sha256_final(ctx, data);
      dealloc();
    }

  return *this;
}

void Hash::dealloc()
{
  if(context)
    {
      sha256_ctx *ctx = (sha256_ctx*)context;
      delete ctx;
      context = NULL;
    }
}

void Hash::clear()
{
  dealloc();
  memset(data, 0, 40);
}

bool Hash::isNull() const
{
  for(int i=0; i<40; i++)
    if(data[i] != 0) return false;
  return true;
}

void Hash::copy(const void* source)
{
  dealloc();
  memcpy(data, source, 40);
}

static char hexDigit(int i)
{
  assert(i >= 0 && i < 16);
  if(i < 10) return '0' + i;
  return 'a' + i - 10;
}

static uint8_t dehexDig(char c)
{
  if(c >= '0' && c <= '9')
    return c - '0';
  if(c >= 'a' && c <= 'f')
    return c - 'a' + 10;
  if(c >= 'A' && c <= 'F')
    return c - 'A' + 10;

  throw std::runtime_error(std::string("Invalid hex character '") + c + "'");
}

static uint8_t dehex(const char *s)
{
  return (dehexDig(*s)<<4) | dehexDig(s[1]);
}

/* Base64 characters. Notice that we use an URL-friendly variant with
   -_ at the end instead of +/. But our decoding routine supports both.
 */
static const char cb64[]="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

// Converts 3 input bytes to 4 output characters
void toBase(const uint8_t *in, char *out)
{
  out[0] = cb64[ in[0] >> 2 ];
  out[1] = cb64[ ((in[0] & 0x03) << 4) | ((in[1] & 0xf0) >> 4) ];
  out[2] = cb64[ ((in[1] & 0x0f) << 2) | ((in[2] & 0xc0) >> 6) ];
  out[3] = cb64[ in[2] & 0x3f ];
}

// Decode character
uint8_t bval(char c)
{
  if(c >= 'A' && c <= 'Z') return c-'A';
  if(c >= 'a' && c <= 'z') return c-'a'+26;
  if(c >= '0' && c <= '9') return c-'0'+52;
  if(c == '-' || c == '+') return 62;
  if(c == '/' || c == '_') return 63;
  if(c == '=') return 0;

  throw std::runtime_error(std::string("Invalid base64 character '") + c + "'");
}

// Converts 4 input characters to 3 output bytes
void fromBase(const char *in, uint8_t *out)
{
  uint8_t v0 = bval(in[0]);
  uint8_t v1 = bval(in[1]);
  uint8_t v2 = bval(in[2]);
  uint8_t v3 = bval(in[3]);
  out[0] = (v0 << 2) | ((v1 & 0x30) >> 4);
  out[1] = ((v1 & 0x0f) << 4) | ((v2 & 0x3c) >> 2);
  out[2] = ((v2 & 0x03) << 6) | v3;
}

static void decodeBase(const std::string &str, uint8_t *out)
{
  assert(str.size() <= 56);

  int pos = 0;
  int outleft = 40;

  while(pos < str.size() && outleft > 0)
    {
      char buf[5] = "AAAA";
      str.copy(buf, 4, pos);

      uint8_t obuf[3];
      fromBase(buf, obuf);

      if(outleft >= 3)
        memcpy(out, obuf, 3);
      else
        memcpy(out, obuf, outleft);

      outleft -= 3;
      out += 3;
      pos += 4;
    }
}

std::string Hash::toBase64() const
{
  if(isNull()) return "00";

  const uint8_t *ptr = data;

  char output[56];

  for(int i=0; i<13; i++)
    toBase(ptr+i*3, output+i*4);
  uint8_t last[3];
  last[0] = ptr[39];
  last[1] = last[2] = 0;
  toBase(last, output+52);

  // Trim trailing As
  int end = 55;
  while(output[end] == 'A' && end>0) end--;

  return std::string(output, end+1);
}

std::string Hash::toHex() const
{
  std::string result;
  result.reserve(80);

  for(int i=0; i<40; i++)
    {
      uint8_t byte = data[i];
      result += hexDigit(byte >> 4);
      result += hexDigit(byte & 0xf);
    }

  return result;
}

std::string Hash::toString() const
{
  return toBase64();
}

void Hash::fromString(const std::string &str)
{
  clear();

  // Special encoding of null hash
  if(str == "00")
    return;

  if(str.size() <= 56)
    {
      decodeBase(str, data);
      return;
    }

  if(str.size() != 80)
    throw std::runtime_error("String is not a valid Hash");

  const char* s = str.c_str();
  for(int i=0; i<40; i++)
    {
      data[i] = dehex(s);
      s += 2;
    }
}

std::ostream& operator<< (std::ostream& out, const Hash &hash)
{
  out << hash.toString();
  return out;
}
