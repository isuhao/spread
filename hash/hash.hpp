#ifndef _SPREAD_HASH_HPP
#define _SPREAD_HASH_HPP

#include <stdint.h>
#include <string>
#include <assert.h>
#include <ostream>
#include <map>

namespace Spread
{
  /* This structure represents an object hash. Most objects in Spread
     are primarily found, indexed and refered to by their hashed
     values.

     A hash is a 40 byte structure, where the first 32 bytes is the
     SHA-256 hash of the object data, and the last 8 bytes represent
     the object's size.

     Hashes can be computed in one go, or iteratively through
     update()/finish().

     Hashes are fully usable as value types, and can safely be copied
     and passed by value. They can also be used as key values in STL
     containers.

     NOTE: Copying partial hashes (where you have called update() but
     not finish()) will work, but make sure the assignment operator=
     or the copy constructor is invoked (ie. NOT a bitwise copy.)
     Otherwise results are undefined.
   */
  struct Hash
  {
    typedef std::map<std::string,Hash> DirMap;

    Hash() : context(NULL) { clear(); }
    Hash(const std::string &hex) : context(NULL) { fromString(hex); }
    Hash(const void *input, uint32_t len) : context(NULL) { hash(input, len); }
    Hash(const Hash &h) : context(NULL) { copy(h); }
    ~Hash() { dealloc(); }

    // Zeros out the hash digest
    void clear();

    // True if the digest is zeroed out
    bool isNull() const;

    // Copy digest verbatim from another source. Must be 40 bytes
    // long.
    void copy(const void* source);

    // Copy contents from another Hash. Will also copy partial hash
    // state (from update()/finish() functions), if any.
    void copy(const Hash &other);

    // Hash a complete buffer, and store the result in this struct.
    void hash(const void *input, uint32_t len);

    /* Perform and finish partial hashing. You can call update()
       sequentially on parts of your buffer. After calling finish(),
       the result will be the same as if you had called hash() on the
       entire buffer.

       There is no 'start' step. Just call update() to begin a new
       hash sequence, and finish() to get the result. The latter
       returns this struct for convenience.

       Calling update() again after finish() starts a new hash
       sequence. Calling finish() with no preceding update() step
       counts as hashing a zero-length buffer.

       Calls to clear(), copy(), hash() and fromString() will also
       abort the sequence.
    */
    void update(const void *input, uint32_t len);
    Hash finish();

    Hash operator=(const Hash &other)
    {
      copy(other);
      return *this;
    }

    // Compare two hashes for equality
    bool operator==(const Hash &other) const
    {
      for(int i=0; i<40; i++)
        if(data[i] != other.data[i])
          return false;
      return true;
    }

    bool operator!=(const Hash &other) const
    { return !(*this == other); }

    // Allows using Hash as a key for std::map
    bool operator<(const Hash &other) const
    {
      for(int i=0; i<40; i++)
        if(data[i] < other.data[i]) return true;
        else if(data[i] > other.data[i]) return false;
      return false;
    }

    // Get and set the size part of the digest
    uint64_t size() const
    {
      return *( (uint64_t*)(data+32) );
    }

    uint64_t &size()
    {
      return *( (uint64_t*)(data+32) );
    }

    // Converts hash digest to compact85 (40-50 characters long)
    std::string toString() const;

    // Converts hash digest to base64url (43-54 characters long)
    std::string toBase64() const;

    // Converts hash digest to an 80 byte long hex string
    std::string toHex() const;

    // Convert string to hash. Accepts both compact85 and hex
    // strings. The string length must represent the exact hash
    // length.
    void fromString(const std::string &str);

    // Return a pointer to the raw binary hash
    const uint8_t *getData() const { return data; }

  private:
    uint8_t data[40];

    // Internal pointer used by update() and finish()
    void *context;

    // Discard all partial hash information created by update(). This
    // has the same effect as calling finish(), except the result is
    // discarded.
    void dealloc();
  };
}

std::ostream& operator<< (std::ostream& out, const Spread::Hash &hash);

#endif
