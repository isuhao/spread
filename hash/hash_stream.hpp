#ifndef __SPREAD_HASH_STREAM_H_
#define __SPREAD_HASH_STREAM_H_

#include "hash.hpp"
#include <mangle/stream/stream.hpp>
#include <mangle/stream/filters/pure_filter.hpp>
#include <mangle/stream/servers/file_stream.hpp>
#include <mangle/stream/servers/outfile_stream.hpp>

/*
  Hash sum wrapping Mangle Stream.

  This filter stream computes the hash sum of all data read or written
  through it. It has one control function finish(), which is
  documented below. The hash sum is computed in the .hash member, and
  also returned by finish(). Otherwise the stream does not affect the
  data flow in any way.

  The class also hosts two static sum() functions, which hash a stream
  (or a file) in one go with no persistant information.

  -- More details: --

  This stream does NOT work well with:
  - pointer streams - only read() and write() calls are summed
  - seeking - unless you want some very special effects, seeking is
              not recommended while hashing.

  For this reason, hasPtr and isSeekable is set to false to prevent
  accidental misuse and hard-to-find bugs. If you need these to be
  true, set them manually.

  Also: The ptr and seek functions will still work even though the
  flags are false, so you can still do seeking or pointer operations
  if you know what you are doing.

  Some possible uses beyond simple stream hashing:

  - partial sum: calculate the hash sum of eg. the first 10k by calling
    read(buf, 10*1024) then finish(). Can be done repeatedly to sum a
    stream in blocks rather than as one sum.

  - concatenation: call setStream() multiple times and read each
    stream in full (or in part) to calculate the sum of several
    streams together into one hash.

  - uni- or bi-directional communication verification. For a socket or
    communication stream used for reading and/or writing data,
    finish() can be called in regular intervals (or at the end) to
    verify that both parties got the same result. It doesn't matter if
    one side's reads are the other's writes, as long as the order is
    the same. Can also be used for signing (eg. through sending a hmac
    of the hash) to verify your identity to the other party.
 */

namespace Spread
{

template <class Hash>
struct HashStreamT : Mangle::Stream::PureFilter
{
  Hash hash;

  // Total number of bytes summed (read + written). It is NEVER reset
  // to zero outside the constructor, so if you want to reset the
  // counter you have to do so manually.
  size_t summed;

  // Start without a stream. Use PureFilter::setStream() before you
  // begin stream operations.
  HashStreamT() : summed(0)
  {
    isSeekable = false;
    hasPtr = false;
  }

  // Start with a stream
  HashStreamT(Mangle::Stream::StreamPtr src) : summed(0)
  {
    setStream(src);
    isSeekable = false;
    hasPtr = false;
  }

  // File version (read or write, depending on second parameter)
  HashStreamT(const std::string &filename, bool write=false) : summed(0)
  {
    using namespace Mangle::Stream;
    Stream *p;
    if(write) p = new OutFileStream(filename);
    else p = new FileStream(filename);
    setStream(StreamPtr(p));
    isSeekable = false;
    hasPtr = false;
  }

  /* Finish and reset the hash summing sequence. The result is
     returned, as well as stored in the .hash member.

     Any read/write operations after finish() restarts the summing
     process, as per update()/finish() in the Hash struct.
   */
  Hash finish()
  {
    return hash.finish();
  }

  size_t read(void *buf, size_t count)
  {
    // Read, then hash
    size_t res = src->read(buf, count);
    // And make sure to only hash actual bytes read
    hash.update(buf, res);
    summed += res;
    return res;
  }

  size_t write(const void *buf, size_t count)
  {
    // Hash, then write
    hash.update(buf, count);
    summed += count;
    return src->write(buf,count);
  }

  // Buffer size used for the sum() functions
  static const int DEFSIZE = 16*1024;

  /* Directly sum THIS stream. See the static sum() function for more
     details.
  */
  Hash sum(int bufsize = DEFSIZE)
  { return sum(*this, bufsize); }

  /* Directly sum an entire stream. The stream is read in increments,
     to avoid loading large files into memory. You can specify the
     buffer size as the second parameter, the default is 16 Kb.

     The algorithm does not depend on size() for non-pointer streams.

     Pointer streams are hashed in one go, and do not advance the file
     pointer.

     Since the position of the file pointer after this operation
     depends on the type of stream given, it is considered undefined.
   */
  static Hash sum(Mangle::Stream::Stream &str, int bufsize = DEFSIZE)
  {
    assert(str.isReadable);

    // Is this a memory stream?
    if(str.hasPtr)
      {
        assert(str.hasSize);

        // Pointer streams makes life easy
        return Hash(str.getPtr(), str.size());
      }

    // No pointers today. Create a buffer and hash in increments.
    char *buf = new char[bufsize];

    Hash result;

    // We do not depend on the stream size, only on eof().
    while(!str.eof())
      {
        size_t num = str.read(buf, bufsize);

        // If we read less than expected, we should be at the
        // end of the stream
        assert(num == bufsize || (num < bufsize && str.eof()));

        // Hash the data
        result.update(buf, num);
      }

    // Clean up and return
    delete[] buf;
    return result.finish();
  }
  static Hash sum(Mangle::Stream::StreamPtr str, int bufsize = DEFSIZE)
  { return sum(*str, bufsize); }

  // Filename version of sum(), through FileStream.
  static Hash sum(const std::string &filename, int buf = DEFSIZE)
  {
    using namespace Mangle::Stream;
    return sum(FileStreamPtr(new FileStream(filename)), buf);
  }
};

typedef HashStreamT<Hash> HashStream;
typedef boost::shared_ptr<HashStream> HashStreamPtr;
}
#endif
