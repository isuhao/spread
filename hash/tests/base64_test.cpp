#include "hash.hpp"

#include <iostream>
using namespace std;

using namespace Spread;

void test(const Hash &h)
{
  Hash h2(h.toString());
  assert(h2 == h);
  cout << "Matched " << h << endl;
}

int main()
{
  Hash h("IC1yZiBDTWFrZ-ZpbGVzLyBDTWFrZU_hY2hlLnR4dCBjbWFrZV9pb");
  cout << "Hash = " << h << endl;
  assert(h == Hash("IC1yZiBDTWFrZ+ZpbGVzLyBDTWFrZU/hY2hlLnR4dCBjbWFrZV9pb"));
  assert(h == Hash("IC1yZiBDTWFrZ-ZpbGVzLyBDTWFrZU/hY2hlLnR4dCBjbWFrZV9pb=="));
  assert(h == Hash("IC1yZiBDTWFrZ-ZpbGVzLyBDTWFrZU/hY2hlLnR4dCBjbWFrZV9pb="));
  assert(h == Hash("IC1yZiBDTWFrZ-ZpbGVzLyBDTWFrZU_hY2hlLnR4dCBjbWFrZV9pbAA"));
  assert(h == Hash("IC1yZiBDTWFrZ-ZpbGVzLyBDTWFrZU_hY2hlLnR4dCBjbWFrZV9pbA="));

  test(h);
  test(Hash("hello",5));
  test(string("________________________________________"));
  test(string("________________________________________AAA===AA=A===A"));
  test(string("______________________________________________________"));
  test(string("____________________________________________________w"));
  test(string("_____________________________________________________w"));
  test(string("_____________________________________________________w=="));
  test(string("AAx"));
  test(string("==x=="));
  test(string("+-+-/_/_-=+/"));
  test(string("__//++--+/-_"));

  return 0;
}
