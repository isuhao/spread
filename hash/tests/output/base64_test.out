Hash = IC1yZiBDTWFrZ-ZpbGVzLyBDTWFrZU_hY2hlLnR4dCBjbWFrZV9pb
Matched IC1yZiBDTWFrZ-ZpbGVzLyBDTWFrZU_hY2hlLnR4dCBjbWFrZV9pb
Matched LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
Matched ________________________________________
Matched ________________________________________
Matched _____________________________________________________w
Matched ____________________________________________________w
Matched _____________________________________________________w
Matched _____________________________________________________w
Matched AAx
Matched AAx
Matched ----____-A-_
Matched ____-----_-_
